modules/SofaMiscTopology/SofaMiscTopology_test/TopologyChecker_test.cpp

```diff
@@ -0,0 +1,1370 @@
+                                {"topology", topoLink}
+            });
+        TopologyChecker* checker = dynamic_cast<TopologyChecker*>(obj.get());
+        checker->init();
+
+        sofa::helper::WriteAccessor< sofa::core::objectmodel::Data<sofa::helper::vector<Topology::Triangle> > > triangles = topoCon->d_triangle;
+
+        // mix triangle
+        triangles[0][0] = triangles[0][1];
+
+        // Topology checher should detect an error
+        EXPECT_MSG_EMIT(Error);
+
+        return checker->checkTriangleTopology() == false;
+    }
+
+    
+    bool testInvalidTriangle2EdgeContainer()
+    {
+        TriangleSetTopologyContainer* topoCon = loadTopo();
+        if (topoCon == nullptr)
+            return false;
+
+        std::string topoLink = "@" + topoCon->getName();
+        auto obj = sofa::simpleapi::createObject(m_topoNode, "TopologyChecker", {
+                                {"name", "checker"},
+                                {"topology", topoLink}
+            });
+        TopologyChecker* checker = dynamic_cast<TopologyChecker*>(obj.get());
+        checker->init();
+
+        sofa::helper::WriteAccessor< sofa::core::objectmodel::Data<sofa::helper::vector<Topology::Triangle> > > triangles = topoCon->d_triangle;
+
+        // Add triangle without updating cross container
+        triangles.push_back(Topology::Triangle(0, 10, 100));
+
+        // container should be ok
+        EXPECT_EQ(checker->checkTriangleContainer(), true);
+
+        // Topology checher should detect an error
+        EXPECT_MSG_EMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), false);
+        // restore good container
+        triangles.resize(triangles.size() - 1);
+        EXPECT_MSG_NOEMIT(Error); // reset no emit error for next true negative test
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), true);
+
+
+        // Mix some triangle vertices
+        int tmp = triangles[0][0];
+        triangles[0][0] = triangles[10][0];
+        triangles[10][0] = tmp;
+        EXPECT_MSG_EMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), false);
+        // restore good container
+        triangles[10][0] = triangles[0][0];
+        triangles[0][0] = tmp;
+        EXPECT_MSG_NOEMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), true);
+
+        // Mix some triangles
+        Topology::Triangle tmpT = triangles[0];
+        triangles[0] = triangles[10];
+        triangles[10] = tmpT;
+        EXPECT_MSG_EMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), false);
+        // restore good container
+        triangles[10] = triangles[0];
+        triangles[0] = tmpT;
+        EXPECT_MSG_NOEMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), true);
+
+        sofa::helper::WriteAccessor< sofa::core::objectmodel::Data<sofa::helper::vector<Topology::Edge> > > edges = topoCon->d_edge;
```

Comment:
Advice: for super long line like that, use auto.
And in this case: `auto edges = sofa::helper::getWriteAccessor(topoCon->d_edge);`

Reply:
thanks, done!